[ian.lock] 지하철 관리자 4단계 PR

Jjungs7 opened the pull request:
## 진행상황

Layered architecture(web - service - repository) 적용
- section, line, station 동일
  - 다른 layer와 데이터를 주고 받을때는 VO만을 활용
  - Controller는 DTO 및 Service만을 이용
  - DAO는 DB와 상호작용하기 위한 최소한의 작업만을 한다

---
벨 안녕하세요! 하다보니 대규모 공사를 한 듯한 느낌이 드네요 😱
피드백 상세하게 해주셔서 감사해요 : ) 하나하나 읽어보면서 의미를 파악하고 적용하려하고 있습니다
너무 많은게 변해서 이게 의도하신게 맞는건지 궁금했어요(이미 너무 멀리 와버린 것 같지만요..). 중간단계까지 리팩토링하고 먼저 피드백 요청드립니다! 이번에도 잘 부탁드립니다 🙏

질문: layered architecture에서 같은 layer에 있는 컴포넌트끼리는 통신해도 괜찮을까요? 순환참조의 원인이 같은 layer에 있는 빈끼리 통신인거같아서요!

추가로 궁금했던 사항은 중간중간에 적어놓았습니다!

Jjungs7 committed:
refactor(tests, line): 테스트 의도 명시, 불필요한 메소드 제거

- 메소드 시그니처만으론 의도를 알아채기 어려운 테스트가 있음
- 불필요한 equals 오버라이딩, 같은 일을 하는 메소드들은 하나의 메소드로 병합

Related:
- https://github.com/next-step/spring-subway-admin-kakao/pull/30/files#r560687891
- https://github.com/next-step/spring-subway-admin-kakao/pull/30/files#r560700714
- https://github.com/next-step/spring-subway-admin-kakao/pull/30/files#r560708305
- https://github.com/next-step/spring-subway-admin-kakao/pull/30/files#r560709836

Jjungs7 committed:
refactor(exception): 상황에 맞게 exception 반환

- 기존에는 try-catch 와 null을 리턴하는 패턴으로 입력 값의 유효성을 처리했음
  - 이로 인해 dao, insert와 같은 곳에서 불필요한 로직이 들어감
- ControllerAdvice를 활용하여 가독성 향상 기대

Related:
- https://github.com/next-step/spring-subway-admin-kakao/pull/30/files#r560693761
- https://github.com/next-step/spring-subway-admin-kakao/pull/30/files#r560705982
- https://github.com/next-step/spring-subway-admin-kakao/pull/30/files?file-filters%5B%5D=.gradle&file-filters%5B%5D=.java&file-filters%5B%5D=.sql&file-filters%5B%5D=.yml&hide-deleted-files=true#r560705982
- https://github.com/next-step/spring-subway-admin-kakao/pull/30/files?file-filters%5B%5D=.gradle&file-filters%5B%5D=.java&file-filters%5B%5D=.sql&file-filters%5B%5D=.yml&hide-deleted-files=true#r560706164

Jjungs7 committed:
style: frontend에도 Id -> ID로 변경

Jjungs7 committed:
refactor(station): Service 생성, dto 및 vo 관리

Repository - Service - Controller
Domain 객체 -    VO   - DTOs

위와 같은 구조로 데이터 흐름이 오가게끔 Section 수정

Related:
- https://github.com/next-step/spring-subway-admin-kakao/pull/30#discussion_r560697983
https://github.com/next-step/spring-subway-admin-kakao/pull/30#discussion_r560691306

Jjungs7 committed:
refactor(line, section): layered architecture 적용

- 다른 layer와 데이터를 주고 받을때는 VO를 활용
- Dao에서 비즈니스 로직을 분리하여 Service로 이동
- Controller는 오직 Service만을 이용하여 데이터 흐름을 제어한다

Related:
- https://github.com/next-step/spring-subway-admin-kakao/pull/30#discussion_r560691306
- https://github.com/next-step/spring-subway-admin-kakao/pull/30#discussion_r560697983
- https://github.com/next-step/spring-subway-admin-kakao/pull/30#discussion_r560699949
- https://github.com/next-step/spring-subway-admin-kakao/pull/30#discussion_r560704413

Jjungs7 commented on src/main/java/subway/line/LineController.java:
```diff
@@ -18,102 +21,76 @@
-                .map(stationDao::findByID)
-                .collect(Collectors.toList()));
+    private List<StationResponse> findStationReponses(LineResultValue lineResultValue) {
+        SectionReadStationsValue sectionReadStationsValue = new SectionReadStationsValue(lineResultValue.getID());
+        StationResultValues stationResultValues = sectionService.findStationsByLineID(sectionReadStationsValue);
```
이것저것 해보다가 결국 LineController에서 LineResponse 따로, List<StationResponse> 따로 불러서 리턴하는 방식으로 회귀한 것 같아요 😭
제일 괜찮다고 생각했던 방식이 Line 에 Sections를 저장하는거였어요. 그러면 데이터 흐름이 아래와 같아지는데
```text
// 'a < b' 는 a 가 b를 포함한다는 뜻으로 사용했습니다

Line < Sections
LineResultValue < SectionResultValues
LineResponse < List<StationResponse> (또는 LineResponses < List<StationResponses>
```

생각해보니 LineResultValue는 LineService에서 계산한 결과를 Controller에 전달해주는거니까 Line과 관련이 없는 Section결과까지 같이 리턴해주는게 맞는지 의문이 들었어요. LineResultValue에 SectionResultValues 결과를 안 넣어주려다보니 데이터가 중간에 사라져서 결국 Controller에서는 접근하지 못하게 되더라구요. 지금 커밋한 방식이 괜찮은지 아니면 LineResultValue 에 SectionResultValue까지 넣는 것으로 생각하고 아예 처음부터 Line에 Section 정보를 저장해도 괜찮을까요?

limmmee-bonjour replied:
제가 정확히 이안의 의도를 잘 파악한건지 모르겠어요! 일단 질문에 대한 저의 답을 드릴게요^^
(혹시 제가 잘못 이해한 부분이 있으면 알려주세요!)
무조건적으로 vo를 만들려고 노력하지 않아도 될 것 같아요.
현재의 Line과 LineResultValue의 차이점이 뭘까요? LineResultValue에서 어떤걸 나타내기 위해 사용한걸까요?
무조건적으로 vo를 만들기보다는 이유가 있는 vo가 만들어져야 할 것 같아요~
그게 아니라면 계속 파싱을 따라가야해서 가독성이 오히려 떨어질 것 같습니다.

현재 구조는 Line -> Section -> Station 정보를 담고 있는 것 같네요!
아마 Line 검색, Section 검색, Station 검색해야할 것 같은데 이런 경우 join을 쓸 수 있고 JPA를 사용하게 된다면 연관관계를 사용한 LazyLoading으로 데이터를 가져올 수 도 있어요^^ (요건 나중에 JPA를 공부하면서 익힐 수 있을거예요!)
그러나 현재는 Line -> Section -> Station 으로 조회해도 될 것 같아요.

제 생각은 현재 구조에서 Line 도메인과 LineResultValue이 별다를게 없으니 말씀해주신대로 SectionResultValue를 포함한 value를 만드는게 더 좋을 것 같습니다 :) 아래와 같은 방식으로 LineService에 만들어도 괜찮을 것 같아요! (물론 답은 아닙니다!)
```
public LineResultValue findBy(LineReadValue readValue){
    Line line = find(readValue);
    return  LineResultValue.of(line, findStationReponses(line));
}

public StationResultValues findStationReponses(Line line) {
    SectionReadStationsValue sectionReadStationsValue = new SectionReadStationsValue(line.getID());
    return sectionService.findStationsByLineID(sectionReadStationsValue);
}
```

단, Line에 Section을 정보를 저장하는 것은 다른 문제입니다. (제가 이해한건 Line 도메인에 Secion을 넣는다는거겠죠?)
(말한대로 JPA 구조에서는 조금 다릅니다. 연관관계에 따라 도메인을 알아서 셋팅해주기때문에..?!)
지금은 mysql 로 데이터를 저장하는 방식이니 Line에 Secion을 주입하려면 DAO에서 다 적용을 해줘야하는데 그렇게 되면 복잡해 질 것 같아요. 이 부분은 고민을 좀 더 해보고 현재는 id 기반으로 저장하는게 좋을 것 같아요!

findStationReponses 메소드는 Controller보다는 Service로 이동하는게 좋을것 같습니다 :)

Jjungs7 commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -44,36 +42,47 @@ public Line insert(Line line) {
+        int affectedRows;
+
         try {
-            return jdbcTemplate.queryForObject(sql, lineRowMapper, id);
-        } catch (DataAccessException e) {
+            affectedRows = jdbcTemplate.update(sql, line.getName(), line.getColor(), line.getID());
+        } catch (DataAccessException ignored) {
+            return null;
+        }
+
+        if (isNotUpdated(affectedRows)) {
             return null;
         }
+
+        return line;
     }
 
-    public boolean update(Long id, LineRequest lineRequest) {
-        String sql = "update line set name = ?, color = ? where id = ?";
-        return jdbcTemplate.update(sql, lineRequest.getName(), lineRequest.getColor(), id) > 0;
+    public Line delete(Line line) {
+        int affectedRows = jdbcTemplate.update("delete from line where id = ?", line.getID());
+
+        if (isNotDeleted(affectedRows)) {
+            return line;
+        }
+
+        return null;
```
db에 저장할 때 잘 됐는지 안 됐는지 확인해야되더라구요. 그래서 아래와 같은 의미로 메소드를 구현했는데 의미가 잘 통할까요?

update 후 똑같은 객체가 돌아왔으면 잘 update 돼서 돌아온 것이다. 실패했으면 실패해서 아무 데이터도 안 온다
delete 후 null 이 돌아왔으면 잘 삭제돼서 데이터가 사라진 것이다. 실패했으면 보낸 객체가 삭제되지 않아서 그대로 돌아온다

limmmee-bonjour replied:
update나 delete로 리턴되는 affectedRows 값은 db에서 몇개가 반영되었는지 그 리턴값을 주는데, 굳이 객체 (Line)를 외부로 노출해서 service에서 그 값을 체킹하게 해야할까요?
JdbcTemplate 은 오래 사용되어진 외부 lib인데 해당 return 값 자체로 반영되었는지 아닌지 체킹하면 될 것 같습니다 :)!

Service, Dao에서 isNotUpdated(isNotDeleted) 의미가 중복되고, Dao에서 확인할 수 있는 문제를 Service에서 까지 보내주는 것 같아요! 이 부분은 고민해 보아요!

limmmee-bonjour commented on src/main/java/subway/exceptions/ExceptionAdvice.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.NoSuchElementException;
+
+@ControllerAdvice
+public class ExceptionAdvice {
```
오호! @ControllerAdvice를 사용하여 Exception을 핸들링 하도록 처리해주셨군요! 👍 💯 
네이밍은 ExceptionAdvice보다는 더 의미를 나타내는 ExceptionHandler가 어때요~?

limmmee-bonjour commented on src/main/java/subway/line/LineController.java:
```diff
@@ -18,102 +21,76 @@
-            return ResponseEntity.badRequest().build();
-        }
+        SectionCreateValue sectionCreateValue = new SectionCreateValue(lineResultValue.getID(),
+                request.getUpStationID(), request.getDownStationID(), request.getDistance());
```
VO를 만들어주셨다면, 파라메터는 LineResultValue, LineRequest 로 받아도 무방할 것 같습니다 :)

limmmee-bonjour commented on src/main/java/subway/section/vo/SectionReadStationsValue.java:
```diff
@@ -0,0 +1,13 @@
+package subway.section.vo;
+
+public class SectionReadStationsValue {
```
의미 있는 value를 만드는 것은 좋아요! 그러나 SectionReadStationsValue 이 value는 뭘 하고 싶은지 어떤 값을 가지고 있는지 (아마 id는 line의 id겠죠?) 전혀 value 만든 목적이라던가 의미를 알 수 없어요! 
어떤 의도로 만들었는지 알 수 있을까요~?

Jjungs7 replied:
"다른 Layer와 통신할 때 무조건 VO로 통신" 이라는 생각으로 작성하다보니 억지로 만든 VO가 다수 생겼네요
이상과 현실의 격차를 생각하며 어느정도 타협하는 것이 좋겠다는 생각이 들었습니다!
위 `SectionReadStationsValue` 는 Section 내의 stationID로 StationResultValue를 만들때 중간 VO 역할을 담당했었는데, 의도가 제대로 드러나지 않고 오히려 가독성을 떨어뜨리는 것 같아 삭제하였습니다

limmmee-bonjour commented on src/main/java/subway/line/vo/LineCreateValue.java:
```diff
@@ -0,0 +1,19 @@
+package subway.line.vo;
+
+public class LineCreateValue {
```
의미있는 value 생성은 좋습니다 👍 👏

limmmee-bonjour commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -44,36 +42,47 @@ public Line insert(Line line) {
-    public boolean update(Long id, LineRequest lineRequest) {
-        String sql = "update line set name = ?, color = ? where id = ?";
-        return jdbcTemplate.update(sql, lineRequest.getName(), lineRequest.getColor(), id) > 0;
+    public Line delete(Line line) {
```
delete의 param이 Line이라 Service에서 불필요한 Line을 만들고 있어요!
dao에서는 굳이 line을 안받아도 될 것 같아요 (제가 말했던 이상향과 현실에서 트레이드오프를 잘 해야할 것 같아요^^;;)
lineId를 받아서 불필요한 Line을 안만든다면 (name, color가 unused 인..? Line이라? 🤔 )
저는 이 방법을 택할 것 같아요!

limmmee-bonjour commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.stream.Collectors;
+
+@Service
+public class LineService {
```
Service 레이어가 만들어졌네용! 책임을 분리하기 굿! 👍 💯 

Q. 질문: layered architecture에서 같은 layer에 있는 컴포넌트끼리는 통신해도 괜찮을까요? 순환참조의 원인이 같은 layer에 있는 빈끼리 통신인거같아서요!
A.같은 layer라면 service -> service를 호출한다는 말씀이시겠죠? 
네 물론 service -> service 도 호출합니다! 그러나 이안이 말씀해주셨던것 처럼 순환참조라 던가 하는 의도치 않은 상황이 발생하기 때문에 항상 고민하고 의존성을 추가해주어야 합니다!

위에서 말한 것 처럼 제가 생각했을 때 Line -> Section -> Station 으로 의존성이 이어져 있는 것 같아서
해당 의존성으로 service가 service를 의존하는 것은 괜찮아 보입니다.
단 StationService에서 LineService를 의존하면 안되겠죠?
의존성이 어떻게 향하는지 항상 고민하고 동일 layer 호출을 해주어야 합니다 😉

limmmee-bonjour commented on src/main/java/subway/line/domain/Line.java:
```diff
@@ -1,39 +1,31 @@
     private final String name;
     private final String color;
 
+    public Line(Long id) {
+        this(id, "unused", "unused");
+    }
+
```
이런 생성자는 필요할까요? 🤔  잘못되어 사용진다면 이슈가 없을까용?

Jjungs7 replied:
억지로 제가 정의한 VO에 맞추려다보니 해괴망측한 코드를 만들었네요 😂

limmmee-bonjour commented on src/main/java/subway/line/vo/LineDeleteValue.java:
```diff
@@ -0,0 +1,13 @@
+package subway.line.vo;
+
+public class LineDeleteValue {
```
LineDeleteValue, LineReadValue 들도 어떤 의미가 있는지, 굳이 사용되어야하는지? 이안의 의도가 궁금합니다!

Jjungs7 replied:
이것도 위에서 언급한 것처럼 제가 정의한 VO틀에 맞추려다보니 나온 VO예요. 의미가 없는 것 같아 삭제했습니다!

limmmee-bonjour commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    // TODO: 날씬해질 수 있을..까?
+    public SectionResultValues create(SectionCreateValue createValue) {
+        if (upAndDownStationsAreEqual(createValue)) {
+            throw new IllegalArgumentException("UpSection and DownSection cannot be equal");
+        }
+
+        Sections sections = sectionDao.findAllSectionsOf(createValue.getLineID());
+        Section upSection = new Section(createValue.getLineID(), createValue.getUpStationID(), 0);
+        Section downSection = new Section(createValue.getLineID(), createValue.getDownStationID(),
+                createValue.getDistanceDiff());
+
+        if (sections.hasNoSections()) {
+            sectionDao.insert(upSection);
+            sectionDao.insert(downSection);
+
+            Sections newSections = sectionDao.findAllSectionsOf(createValue.getLineID());
+            return newSections.toResultValues();
+        }
+
+        Section newSection = sections.createSection(upSection, downSection);
+        if (newSection == null) {
+            throw new IllegalArgumentException("Cannot create section. Distance exceeds existing section distance");
+        }
+
+        sectionDao.insert(newSection);
+        Sections newSections = sectionDao.findAllSectionsOf(createValue.getLineID());
+        return newSections.toResultValues();
+    }
```
자! 이안 ㅎㅎ TODO를 달아주셨네요! 👍 
[도전!] 너무 많은 일을 하고 있는 것 같죠? 한번 날씬하게 다이어트를 시켜줍시다!ㅎㅎ
뭔가 어렵다면, 고민이 된다면 어떤 것이 고민이 되는지 DM 주세요 화이팅! 💪

limmmee-bonjour commented on src/main/java/subway/line/LineController.java:
```diff
@@ -18,102 +21,76 @@
     }
 
     @PostMapping
     public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest request) {
```
Controller에서 Line과 Section을 생성하고 있네요! 
만약 Line은 생성되었는데 Section 생성할 때 DB응답이 없어 에러가 났다면! 
내가 의도했던건 Line,Section 두개다가 모두 생성되거나, 둘다 생성되지 않는것(에러가 있다면) 이라면 
현재 코드에서는 의도사항을 반영한걸까요?
작업 단위를 묶어 DB 반영을 일관적이게 싶다면 어떻게 해야할까요? 고민해 봅시다 :)
+) 힌트는?! 토비의 스프링 6장..?!ㅎㅎㅎㅎㅎㅎ

limmmee-bonjour reviewed (changes requested):
안녕하세요 이안! 벨입니다 😄 
짧은 시간안에 대규모 리팩토링을 해주셨어요 👍  고생하셨어요! 👏 
고민하신 흔적이 많이 보입니다!
그러나 제가 드린 코멘트중 value 관련된 부분에 대해 다시 들여다 볼 필요가 있는 것 같아요.
너무 많은 불필요한 value들이 생성되어서 오히려 가독성이 떨어지는 것 같습니다.
무조건적으로 create/result value를 만들 필요는 전혀 없습니다. 그 value를 만든 납득할만한 이유가 중요한 것 같아요. 이 부분 한번 확인 부탁드려요 :)
또 질문해 주신 부분에 대해 코멘트 드렸으니 확인해보시고, 궁금하신 점이 있으시면 DM 주셔요!
그럼 한번 더 달려봅시당! 화이팅이예요! 💪

Jjungs7 committed:
refactor(DTO): Jackson parsing 방법 변경

parameter명이 변경되면 개발자가 직접 ConstructorProperties의 값들을 변경해야했음
JsonCreator를 활용하여 불필요한 작업 최소화

*주의* IntelliJ 설정 변경이 필요함
Settings -> Build, Execution, Deployment -> Compiler -> Java Compiler
  Additional command line parameters:
  -parameters
를 추가해야 정상적으로 작동함

Jjungs7 commented on src/main/java/subway/line/dto/LineRequest.java:
```diff
@@ -10,7 +10,7 @@
     private int extraFare;
 
-    @ConstructorProperties({"name", "color", "upStationID", "downStationID", "distance"})
+    @JsonCreator
```
> ConstructorProperties 를 사용하신 이유가 있을까요? (Jackson 역직렬화 할 때 기본생성자 없이 사용하려는 목적일까요?)
> 그렇지만 이런경우 name이나, color 네이밍이 변경된다면, 변경이 누락될 가능성도 있을것 같은데 혹은 스펠링이 틀렸다던가..?
> 어떻게 생각하세요?

오랜 삽질 끝에 결국 @JsonCreator를 쓰게 되었어요! 사실 ParameterNamesModule 를 적용해서 JsonCreator조차 쓰지 않으려했는데(JsonCreator는 Jackson 의존성이라서 다른 parser로 바꾸면 JsonCreator 전체 다 바꿔야한다고 해서요) 아직 configuration이 익숙하지 않은 탓인지 제대로 등록 안 되는 것 같더라구요 😭

ConstructorProperties는 다른 파서가 해당 어노테이션을 지원하면 코드를 바꾸지 않아도 된다길래 사용했습니다! 하지만 새로운 클래스를 만들때마다 직접 작성해야 한다는 번거로움이 앞에서 언급한 편리함보다 더 큰 것 같네요...

Jjungs7 committed:
refactor: 불필요한 VO 제거 / Dao, service을 날씬하게

- SectionService에 알맞지 않던 로직 Sections로 위임
- Dao에서 확인하던 조건을 Service 또는 Domain으로 위임

Jjungs7 committed:
refactor(service): try문 분리하여 메소드가 한 가지 일만 더 잘하게 수정

- 기존 메소드는 에러 처리와 repository 통신 두 가지 일을 하였음

Jjungs7 committed:
refactor: 결과 변환 방식변경

Jjungs7 commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,75 @@
+        Section downSection = new Section(createValue.getLineID(), createValue.getDownStationID(),
+                createValue.getDistanceDiff());
+
+        if (sections.hasNoSections()) {
+            sectionDao.insert(upSection);
+            sectionDao.insert(downSection);
+
+            return;
+        }
```
> 자! 이안 ㅎㅎ TODO를 달아주셨네요! 👍
> [도전!] 너무 많은 일을 하고 있는 것 같죠? 한번 날씬하게 다이어트를 시켜줍시다!ㅎㅎ
> 뭔가 어렵다면, 고민이 된다면 어떤 것이 고민이 되는지 DM 주세요 화이팅! 💪

제가 이 부분까지 모두 `sections.insertAndGetNewSection()` 에 위임하고싶었는데 잘 안 되더라구요. 먼저 제가 생각한 두 가지 방법을 모두 적어볼게요✍️
- controller 입장에선 [요청1: LineController.createLine()](https://github.com/next-step/spring-subway-admin-kakao/pull/46/files#diff-859bb950d5635a3d5a88eb92316552a54dca5a17a82ceaa8f3675e7d0faa27e1R33)이나 [요청2: LineController.addSection()](https://github.com/next-step/spring-subway-admin-kakao/pull/46/files#diff-859bb950d5635a3d5a88eb92316552a54dca5a17a82ceaa8f3675e7d0faa27e1R33)에서의 구간생성 작업은 동일하다. 따라서 `sectionService.새로운_노선을_위해_상행과_하행_구간_추가()` (상행/하행에 대한 유효성 검사만 하고 두 구간을 바로 추가하는 메소드)와 `sectionService.예외처리_없는_create()` (위 메소드에서 if문을 제외한 것) 두 개의 메소드로 나누는 것은 controller에게 domain에 대한 정보를 노출하는 것이다. 따라서 현재처럼 `create()`에서 노선 추가에 대한 경우만 예외처리하는 것이 좋다
- 새로운 노선이 추가되면 상행/하행 구간 모두 추가되는 작업(요청1)은 기존 노선에 새로운 구간을 추가하는 작업(요청2)과는 별개이므로 두 메소드를 분리하는 것이 좋다

언뜻보면 현재 create() 메소드가 너무 많은 일을 담당하는 것 같아요(유효성검사, 예외처리, 구간추가). 노선 생성에 대한 케이스를 따로 메소드로 빼면 좀 더 날씬해질 것 같은데 괜찮을까요?

limmmee-bonjour replied:
제가 잘 이해한건지 모르겠어요 :) 혹시 제가 잘못 이해한 부분이 있다면 알려주세요!
메소드가 길다면 일단 메소드를 아예 분리하기 보다는 포함되는 부분을 메소드로 만들어볼 수 있을 것 같아요~
말씀하신대로 createLine, addSection이 동일한 일을 하기 때문에 메소드를 분리하는 것보다는
긴 메소드를 가독성이 좀 더 있게 바꿔보는게 좋을 것 같아요~

upAndDownStationsAreEqual 은 여기서만 사용한다면, valide형태로 바꿔서 if문, exception 처리부분을 포함한 validUpAndDownStationsAreEqual 정도로 뺄 수 있을 것 같습니다.

신규로 추가가(2개의 section 셋팅) 필요하다면 메소드를 분리하여 요런식으로 좀 더 간결하게 메소드를 만들 수 있지 않을까요~?
(물론 네이밍은 의도가 잘 드러나도록 수정해주셔야해요 :)) 한번 고민해주세요 !
```
public void create(SectionCreateValue createValue) {
    validUpAndDownStationsAreEqual(createValue);

    Sections sections = sectionDao.findAllSectionsOf(createValue.getLineID());
    if (sections.hasNoSections()) {
        init(createValue);
        return;
    }

    Section newSection = sections.getNewSectionIfValid(createValue);
    sectionDao.insert(newSection);
}

public void init(SectionCreateValue createValue){
    Section upSection = new Section(createValue.getLineID(), createValue.getUpStationID(), 0);
    Section downSection = new Section(createValue.getLineID(), createValue.getDownStationID(),
                                      createValue.getDistanceDiff());

    sectionDao.insert(upSection);
    sectionDao.insert(downSection);
}
```

Jjungs7 committed:
refactor(sections): 많은 일을 담당하던 메소드의 일 분배

- Sections 컬렉션은 모든 구간에 대하여 lineID 가 유일해야함

Jjungs7 committed:
fix(노선): 구간이 잘못되어도 노선이 생성되지 않음

- 단위 테스트 추가

limmmee-bonjour commented on src/main/java/subway/exceptions/ExceptionHandlers.java:
```diff
@@ -8,7 +8,7 @@
 
 @ControllerAdvice
-public class ExceptionAdvice {
+public class ExceptionHandlers {
```
네이밍 변경 좋아요 👍

limmmee-bonjour commented on src/main/java/subway/line/LineController.java:
```diff
@@ -54,43 +50,39 @@ public LineController(LineService lineService, SectionService sectionService) {
 
     @PutMapping("/{lineID}")
     public ResponseEntity<LineResponse> updateLine(@PathVariable Long lineID, @RequestBody LineRequest lineRequest) {
-        LineUpdateValue lineUpdateValue = new LineUpdateValue(lineID, lineRequest.getName(), lineRequest.getColor());
```
불필요한 value는 삭제해 코드도 간결해지고 가독성이 더 좋아졌어요 👍

limmmee-bonjour commented on src/main/java/subway/line/LineService.java:
```diff
@@ -34,39 +40,45 @@ public LineResultValue create(LineCreateValue createValue) {
+        } catch (DataAccessException e) {
+            throw new NoSuchElementException(
+                    String.format("%s\nCould not retrieve line with id: %d",
+                            e.getMessage(), lineID));
```
해당 에러를 꼭 service에서 잡아야 할까요? DataAccessException 관련 에러는 DAO에서 제어해야할 Exception인 것 같은데 어떻게 생각하세요~?
DataAccessException은 이전처럼 Dao에 있어야할 것 같아요! Dao에서 의도치 않은 상태라면 Exception으로 처리하는건 어떻게 생각해요~?

limmmee-bonjour commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -24,8 +22,7 @@ public SectionService(SectionDao sectionDao, StationDao stationDao) {
 
-    // TODO: 날씬해질 수 있을..까?
-    public SectionResultValues create(SectionCreateValue createValue) {
+    public void create(SectionCreateValue createValue) {
```
다음 프로젝트에서는 [이 글](https://goddaehee.tistory.com/167)을 읽어보시고 적용해보아요 :)

limmmee-bonjour reviewed (approved):
안녕하세요 이안! 벨입니다 :)
vo를 정리해주셔서 훨씬 가독성이 높아지고 깔끔해졌습니다. 제가 의견드린 부분 잘 반영해주셨어요 👍 
질문주신 부분에 대해 코멘트 드렸는데, 혹시 제가 잘못 이해한 부분이 있다면 알려주세요 :)
몇가지 의견 드렸으니 다음 프로젝트 진행시 반영해주시면 좋을 것 같아요!
지하철 노선도 구현하시느라 고생 많으셨습니다. 궁금하신점이 있으시다면 언제든지 DM 주셔요!
그럼 화이팅입니다 😄
